internal/rbd/controllerserver.go

```diff
@@ -158,6 +152,39 @@ func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.
 	return rbdVol, nil
 }
 
+// getVolumeFromID gets the rbd image details from the volumeID and also takes
+// volume lock and creates the new user credential.
+// we need to call cs.cleanup() in defer to remove the lock and credentials.
+func (cs *ControllerServer) getVolumeFromID(ctx context.Context, volumeID string, secrets map[string]string) (*rbdVolume, *util.Credentials, error) {
+	// validate the volume ID
+	cr, err := util.NewUserCredentials(secrets)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if acquired := cs.VolumeLocks.TryAcquire(volumeID); !acquired {
+		return nil, cr, util.ErrOperationExists
+	}
+
+	rbdVol, err := genVolFromVolID(ctx, volumeID, cr, secrets)
```

Comment:
@Madhu-1 why dont we unlock or cleanup from this function if we are done with decoding ?

Reply:
the lock needs to exist till the last part of the RPC calls. unlocking or cleanup from this function is not possible

Reply:
Lock and unlock from either caller or callee is what I am suggesting.

Reply:
This was done to avoid a lot of code duplication which was suggested as an enhancement in the last PR. if required i can change it back. Please let me know  your suggestion

Reply:
@nixpanic what do you think?

Reply:
I prefer to have locks taken and released in the same function. It should be an exception where this is not done.

Maybe introduce a functions like:
```golang
func (cs *ControllerServer) lockVolume(volumeID string) error {
    ...
}

func (cs *ControllerServer) unlockVolume(volumeID string) error {
    ...
}
```

This makes it really easy to re-use the functions everywhere, and set a 
```golang
defer cs.unlockVolume(volID)
```
after obtaining the lock.

I understand/guess that the `cleanup()` function was introduced to reduce the code in certain functions. However, the function moves away from being clear on when/what actions are needed, making maintenance and understanding of the code harder.

If reducing the complexity/size of a function is needed, it is better to introduce (small) helper functions that handle part of the functions work.

Reply:
already lock and unlock are smaller functions, introducing one more wrapper around it is not a good idea. I removed the commit which does this enhancement for now for clean code maintenance and reading.